File: Delphi-source-code/MyInventory/MyInventoryMenu/src/uInventoryMenuBL.pas
unit uInventoryMenuBL;

interface

uses
  System.SysUtils;

type
  TMenuCostItem = record
    MenuItemId: Integer;
    ItemName: string;
    Category: string;
    FoodCost: Currency;
    LaborCost: Currency;
    OverheadCost: Currency;
    TotalCost: Currency;
    SellingPrice: Currency;
    ProfitMargin: Double;
  end;

  TMenuCategorySummary = record
    Category: string;
    ItemCount: Integer;
    AvgPrice: Currency;
    AvgCost: Currency;
    AvgMargin: Currency;
  end;

  TTableAvailability = record
    TableId: Integer;
    TableNumber: Integer;
    Capacity: Integer;
    Zone: string;
    Status: string;
    IsAvailable: Boolean;
  end;

  TInventoryMenuBL = class
  public
    function GetMenuCosts: TArray<TMenuCostItem>;
    function GetMenuCategorySummary: TArray<TMenuCategorySummary>;
    function GetTableAvailability: TArray<TTableAvailability>;
  end;

const
  SQL_GET_MENU_COSTS =
    'SELECT mi.MenuItemId, mi.ItemName, mi.Category, ' +
    '       mi.FoodCost, mi.LaborCost, mi.OverheadCost, ' +
    '       (mi.FoodCost + mi.LaborCost + mi.OverheadCost) AS TotalCost, ' +
    '       mi.SellingPrice, ' +
    '       ((mi.SellingPrice - (mi.FoodCost + mi.LaborCost + mi.OverheadCost)) / mi.SellingPrice * 100) AS ProfitMargin ' +
    'FROM MenuItems mi ' +
    'ORDER BY mi.Category, mi.ItemName';

  SQL_GET_MENU_CATEGORY_SUMMARY =
    'SELECT mi.Category, COUNT(*) AS ItemCount, ' +
    '       AVG(mi.SellingPrice) AS AvgPrice, ' +
    '       AVG(mi.FoodCost + mi.LaborCost + mi.OverheadCost) AS AvgCost, ' +
    '       AVG((mi.SellingPrice - (mi.FoodCost + mi.LaborCost + mi.OverheadCost)) / mi.SellingPrice * 100) AS AvgMargin ' +
    'FROM MenuItems mi ' +
    'GROUP BY mi.Category ' +
    'ORDER BY mi.Category';

  SQL_GET_TABLE_AVAILABILITY =
    'SELECT t.TableId, t.TableNumber, t.Capacity, t.Zone, t.Status, ' +
    '       CASE WHEN t.Status = ''Available'' THEN 1 ELSE 0 END AS IsAvailable ' +
    'FROM Tables t ' +
    'ORDER BY t.Zone, t.TableNumber';

implementation

{ TInventoryMenuBL }

function TInventoryMenuBL.GetMenuCosts: TArray<TMenuCostItem>;

  procedure AddItem(var Arr: TArray<TMenuCostItem>; AId: Integer;
    const AName, ACategory: string; AFood, ALabor, AOverhead, APrice: Currency);
  var
    Idx: Integer;
    Total: Currency;
  begin
    Idx := Length(Arr);
    SetLength(Arr, Idx + 1);
    Total := AFood + ALabor + AOverhead;
    Arr[Idx].MenuItemId := AId;
    Arr[Idx].ItemName := AName;
    Arr[Idx].Category := ACategory;
    Arr[Idx].FoodCost := AFood;
    Arr[Idx].LaborCost := ALabor;
    Arr[Idx].OverheadCost := AOverhead;
    Arr[Idx].TotalCost := Total;
    Arr[Idx].SellingPrice := APrice;
    if APrice > 0 then
      Arr[Idx].ProfitMargin := ((APrice - Total) / APrice) * 100
    else
      Arr[Idx].ProfitMargin := 0;
  end;

begin
  Result := nil;
  AddItem(Result, 1,  'Caesar Salad',      'Appetizer',    3.75, 1.50, 0.80, 12.50);
  AddItem(Result, 2,  'Bruschetta',        'Appetizer',    2.40, 1.20, 0.60, 9.50);
  AddItem(Result, 3,  'Soup of the Day',   'Appetizer',    2.10, 1.00, 0.50, 8.00);
  AddItem(Result, 4,  'Grilled Salmon',    'Main Course',  8.50, 3.00, 1.50, 24.00);
  AddItem(Result, 5,  'Margherita Pizza',  'Main Course',  4.20, 2.00, 1.00, 16.00);
  AddItem(Result, 6,  'Chicken Parmesan',  'Main Course',  6.80, 2.50, 1.20, 19.50);
  AddItem(Result, 7,  'Risotto',           'Main Course',  5.10, 2.20, 1.00, 18.00);
  AddItem(Result, 8,  'Seafood Pasta',     'Main Course',  9.20, 3.00, 1.50, 22.00);
  AddItem(Result, 9,  'Tiramisu',          'Dessert',      3.20, 1.00, 0.50, 10.00);
  AddItem(Result, 10, 'Panna Cotta',       'Dessert',      2.80, 0.80, 0.40, 9.00);
  AddItem(Result, 11, 'Lemonade',          'Beverage',     0.85, 0.50, 0.30, 5.00);
  AddItem(Result, 12, 'Espresso',          'Beverage',     0.60, 0.40, 0.25, 4.00);
end;

function TInventoryMenuBL.GetMenuCategorySummary: TArray<TMenuCategorySummary>;

  procedure AddCategory(var Arr: TArray<TMenuCategorySummary>;
    const ACategory: string; ACount: Integer; AAvgPrice, AAvgCost, AAvgMargin: Currency);
  var
    Idx: Integer;
  begin
    Idx := Length(Arr);
    SetLength(Arr, Idx + 1);
    Arr[Idx].Category := ACategory;
    Arr[Idx].ItemCount := ACount;
    Arr[Idx].AvgPrice := AAvgPrice;
    Arr[Idx].AvgCost := AAvgCost;
    Arr[Idx].AvgMargin := AAvgMargin;
  end;

begin
  Result := nil;
  AddCategory(Result, 'Appetizer',   3, 10.00,  3.45, 65.50);
  AddCategory(Result, 'Main Course', 5, 19.90,  9.40, 52.76);
  AddCategory(Result, 'Dessert',     2,  9.50,  2.93, 69.16);
  AddCategory(Result, 'Beverage',    2,  4.50,  0.97, 78.44);
end;

function TInventoryMenuBL.GetTableAvailability: TArray<TTableAvailability>;

  procedure AddTable(var Arr: TArray<TTableAvailability>; AId, ANumber,
    ACapacity: Integer; const AZone, AStatus: string);
  var
    Idx: Integer;
  begin
    Idx := Length(Arr);
    SetLength(Arr, Idx + 1);
    Arr[Idx].TableId := AId;
    Arr[Idx].TableNumber := ANumber;
    Arr[Idx].Capacity := ACapacity;
    Arr[Idx].Zone := AZone;
    Arr[Idx].Status := AStatus;
    Arr[Idx].IsAvailable := (AStatus = 'Available');
  end;

begin
  Result := nil;
  AddTable(Result, 1,  1,  2, 'Indoor',   'Occupied');
  AddTable(Result, 2,  2,  2, 'Indoor',   'Available');
  AddTable(Result, 3,  3,  4, 'Indoor',   'Occupied');
  AddTable(Result, 4,  4,  4, 'Indoor',   'Reserved');
  AddTable(Result, 5,  5,  6, 'Indoor',   'Available');
  AddTable(Result, 6,  6,  2, 'Bar',      'Occupied');
  AddTable(Result, 7,  7,  2, 'Bar',      'Available');
  AddTable(Result, 8,  8,  4, 'Bar',      'Occupied');
  AddTable(Result, 9,  9,  4, 'Patio',    'Available');
  AddTable(Result, 10, 10, 6, 'Patio',    'Occupied');
  AddTable(Result, 11, 11, 8, 'Patio',    'Available');
  AddTable(Result, 12, 12, 2, 'Patio',    'Reserved');
  AddTable(Result, 13, 13, 4, 'Private',  'Available');
  AddTable(Result, 14, 14, 8, 'Private',  'Reserved');
  AddTable(Result, 15, 15, 10,'Private',  'Available');
end;

end.

File: MyInventoryMenu.Core/Application/Services/MyInventoryMenu/uInventoryMenuBL.cs
using System.Globalization;
using Microsoft.Extensions.Logging;
using MyInventoryMenu.Core.Application.DTOs.Core;

namespace MyInventoryMenu.Core.Application.Services.MyInventoryMenu;

public class uInventoryMenuBL : IuInventoryMenuBL
{
    private readonly ILogger<uInventoryMenuBL> _logger;

    public uInventoryMenuBL(ILogger<uInventoryMenuBL> logger)
    {
        _logger = logger;
    }

    /// <summary>
    /// Synchronous version ported from TInventoryMenuBL.GetMenuCosts.
    /// </summary>
    public List<TMenuCostItem> GetMenuCosts()
    {
        var result = new List<TMenuCostItem>(12);

        AddMenuItem(result, 1,  "Caesar Salad",     "Appetizer",   3.75m, 1.50m, 0.80m, 12.50m);
        AddMenuItem(result, 2,  "Bruschetta",       "Appetizer",   2.40m, 1.20m, 0.60m,  9.50m);
        AddMenuItem(result, 3,  "Soup of the Day",  "Appetizer",   2.10m, 1.00m, 0.50m,  8.00m);
        AddMenuItem(result, 4,  "Grilled Salmon",   "Main Course", 8.50m, 3.00m, 1.50m, 24.00m);
        AddMenuItem(result, 5,  "Margherita Pizza", "Main Course", 4.20m, 2.00m, 1.00m, 16.00m);
        AddMenuItem(result, 6,  "Chicken Parmesan", "Main Course", 6.80m, 2.50m, 1.20m, 19.50m);
        AddMenuItem(result, 7,  "Risotto",          "Main Course", 5.10m, 2.20m, 1.00m, 18.00m);
        AddMenuItem(result, 8,  "Seafood Pasta",    "Main Course", 9.20m, 3.00m, 1.50m, 22.00m);
        AddMenuItem(result, 9,  "Tiramisu",         "Dessert",     3.20m, 1.00m, 0.50m, 10.00m);
        AddMenuItem(result, 10, "Panna Cotta",      "Dessert",     2.80m, 0.80m, 0.40m,  9.00m);
        AddMenuItem(result, 11, "Lemonade",         "Beverage",    0.85m, 0.50m, 0.30m,  5.00m);
        AddMenuItem(result, 12, "Espresso",         "Beverage",    0.60m, 0.40m, 0.25m,  4.00m);

        _logger.LogDebug("GetMenuCosts returned {Count} items", result.Count);
        return result;
    }

    /// <summary>
    /// Synchronous version ported from TInventoryMenuBL.GetMenuCategorySummary.
    /// </summary>
    public List<TMenuCategorySummary> GetMenuCategorySummary()
    {
        var result = new List<TMenuCategorySummary>(4);

        AddCategory(result, "Appetizer",   3, 10.00m,  3.45m, 65.50m);
        AddCategory(result, "Main Course", 5, 19.90m,  9.40m, 52.76m);
        AddCategory(result, "Dessert",     2,  9.50m,  2.93m, 69.16m);
        AddCategory(result, "Beverage",    2,  4.50m,  0.97m, 78.44m);

        _logger.LogDebug("GetMenuCategorySummary returned {Count} categories", result.Count);
        return result;
    }

    /// <summary>
    /// Synchronous version ported from TInventoryMenuBL.GetTableAvailability.
    /// </summary>
    public List<TTableAvailability> GetTableAvailability()
    {
        var result = new List<TTableAvailability>(15);

        AddTable(result, 1,   1,  2,  "Indoor",  "Occupied");
        AddTable(result, 2,   2,  2,  "Indoor",  "Available");
        AddTable(result, 3,   3,  4,  "Indoor",  "Occupied");
        AddTable(result, 4,   4,  4,  "Indoor",  "Reserved");
        AddTable(result, 5,   5,  6,  "Indoor",  "Available");
        AddTable(result, 6,   6,  2,  "Bar",     "Occupied");
        AddTable(result, 7,   7,  2,  "Bar",     "Available");
        AddTable(result, 8,   8,  4,  "Bar",     "Occupied");
        AddTable(result, 9,   9,  4,  "Patio",   "Available");
        AddTable(result, 10, 10,  6,  "Patio",   "Occupied");
        AddTable(result, 11, 11,  8,  "Patio",   "Available");
        AddTable(result, 12, 12,  2,  "Patio",   "Reserved");
        AddTable(result, 13, 13,  4,  "Private", "Available");
        AddTable(result, 14, 14,  8,  "Private", "Reserved");
        AddTable(result, 15, 15, 10,  "Private", "Available");

        _logger.LogDebug("GetTableAvailability returned {Count} tables", result.Count);
        return result;
    }

    /// <summary>
    /// Async version ported from TInventoryMenuBL.GetMenuCosts — builds the full menu cost breakdown
    /// with computed TotalCost and ProfitMargin for each item.
    /// </summary>
    public async Task<TMenuCostItem[]> GetMenuCostsAsync(CancellationToken ct = default)
    {
        await Task.CompletedTask;
        return GetMenuCosts().ToArray();
    }

    /// <summary>
    /// Async version ported from TInventoryMenuBL.GetMenuCategorySummary — returns pre-computed
    /// category-level aggregates (count, avg price, avg cost, avg margin).
    /// </summary>
    public async Task<TMenuCategorySummary[]> GetMenuCategorySummaryAsync(CancellationToken ct = default)
    {
        await Task.CompletedTask;
        return GetMenuCategorySummary().ToArray();
    }

    /// <summary>
    /// Async version ported from TInventoryMenuBL.GetTableAvailability — returns the full table layout
    /// with zone, status, and computed IsAvailable flag.
    /// </summary>
    public async Task<TTableAvailability[]> GetTableAvailabilityAsync(CancellationToken ct = default)
    {
        await Task.CompletedTask;
        return GetTableAvailability().ToArray();
    }

    /// <summary>
    /// Ported from the nested AddItem procedure inside TInventoryMenuBL.GetMenuCosts.
    /// Computes TotalCost = FoodCost + LaborCost + OverheadCost and
    /// ProfitMargin = ((SellingPrice - TotalCost) / SellingPrice) * 100 when SellingPrice > 0.
    /// </summary>
    private static void AddMenuItem(
        List<TMenuCostItem> list,
        int menuItemId,
        string itemName,
        string category,
        decimal foodCost,
        decimal laborCost,
        decimal overheadCost,
        decimal sellingPrice)
    {
        decimal totalCost = foodCost + laborCost + overheadCost;

        decimal profitMargin;
        if (sellingPrice > 0m)
        {
            profitMargin = Math.Round((sellingPrice - totalCost) / sellingPrice * 100m, 2, MidpointRounding.ToEven);
        }
        else
        {
            profitMargin = 0m;
        }

        list.Add(new TMenuCostItem
        {
            MenuItemId = menuItemId,
            ItemName = itemName,
            Category = category,
            FoodCost = foodCost,
            LaborCost = laborCost,
            OverheadCost = overheadCost,
            TotalCost = totalCost,
            SellingPrice = sellingPrice,
            ProfitMargin = (double)(profitMargin)
        });
    }

    /// <summary>
    /// Ported from the nested AddCategory procedure inside TInventoryMenuBL.GetMenuCategorySummary.
    /// </summary>
    private static void AddCategory(
        List<TMenuCategorySummary> list,
        string category,
        int itemCount,
        decimal avgPrice,
        decimal avgCost,
        decimal avgMargin)
    {
        list.Add(new TMenuCategorySummary
        {
            Category = category,
            ItemCount = itemCount,
            AvgPrice = avgPrice,
            AvgCost = avgCost,
            AvgMargin = avgMargin
        });
    }

    /// <summary>
    /// Ported from the nested AddTable procedure inside TInventoryMenuBL.GetTableAvailability.
    /// IsAvailable is computed as (Status == "Available"), matching the legacy check (AStatus = 'Available').
    /// </summary>
    private static void AddTable(
        List<TTableAvailability> list,
        int tableId,
        int tableNumber,
        int capacity,
        string zone,
        string status)
    {
        list.Add(new TTableAvailability
        {
            TableId = tableId,
            TableNumber = tableNumber,
            Capacity = capacity,
            Zone = zone,
            Status = status,
            IsAvailable = string.Equals(status, "Available", StringComparison.Ordinal)
        });
    }
}
